### ISSUE TITLE
no-useless-escape and curly braces

### ISSUE THREAD

**Issue by silverwind** [eslint/eslint#8553]
**Tell us about your environment**

* **ESLint Version:** 3.19.0
* **Node Version:** 7.10.0
* **npm Version:** 4.5.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

```yaml
rules:
  no-useless-escape: [2]
````

**What did you do? Please include the actual source code causing the issue.**

```js
const match = /\{\}/.test('{}');
```

**What did you expect to happen?**

`no-useless-escape` should find and warn on the useless `\{` and `\}` escape sequences in the RegExp. The only case where curly braces need to be escaped is when the value between an opening and closing bracket matches `^[0-9]+,?[0-9]*$`, otherwise the escaping is useless. Also see the [spec](https://tc39.github.io/ecma262/#prod-QuantifierPrefix).

**What actually happened? Please include the actual, raw output from ESLint.**

The code passes.

**Comment by ilyavolodin** [comment]
Sounds like a bug to me. We had to relax this rule a few times, since it was reporting false positives, but since this is defined in the spec, feels like we should support it. @not-an-aardvark thoughts?

**Comment by not-an-aardvark** [comment]
This is working as intended and a duplicate of https://github.com/eslint/eslint/issues/7772.

**Comment by silverwind** [comment]
Thanks for the link, https://github.com/eslint/eslint/issues/7656#issuecomment-267878150 covers this case.

I wonder if it might be wise to allow the user to specify target engines in the configuration, say for example [`v8`, `spidermonkey`, `chakra`], which would then cause the Annex B rules to be applied on `no-useless-escape` and possibly other rules, if all engines support it.

**Comment by silverwind** [comment]
I'll close this, but I'm still wondering if I shall open a separate issue for a possible `engines` configuration option. Is it worth evaluating?

**Comment by ilyavolodin** [comment]
@silverwind We usually try to avoid doing things like that, due to the fact that it puts a lot of maintenance burden on the team. It also leads to length and not always pleasant discussions about why this engine was included, but this other one was not. We prefer things like that to be done in plugins.
